Preserve preprocessing passpipeline attribute during `WrapEntryPointsPass`.

MaheshRavishankar opened the pull request:
The wrap ABI pass moves the body of the entry point function to a different function. The attribute needs to be moved to the newly created function to work as expected.

Towards #20811 

Depends on #20832

MaheshRavishankar committed:
Preserve preprocessing passpipeline attribute during `WrapEntryPointsPass`

The wrap ABI pass moves the body of the entry point function to a
different function. The attribute needs to be moved to the newly
created function to work as expected.

Signed-off-by: MaheshRavishankar <mravisha@amd.com>

benvanik reviewed (changes requested):
This feels like too much goo/coupling for a single attribute copy. I'll take a deeper look in the morning for suggestions, but util should have no concept of a preprocessing pipeline.

benvanik reviewed (changes requested):
Erg, just now reading that PreprocessingPassPipelineAttr is in util - that's not good.

It should be renamed to PassPipelineAttr as it has nothing to do with preprocessing.

The AttrBasedPipelinePass should also be moved to util (as it's generally useful) and can take as an option the attribute name it's looking for (`createAttrBasedPipelinePass("preprocessing")`).

That the attribute name is not dialect based is shady as generally you can't add attributes to ops you don't control without scoping them - `util.pipelines = {"preprocessing": #util.pass_pipeline<....>}` is how you'd handle it safely to avoid conflicts and allow multiple pipelines to be specified to run at different types. Each time the pass ran for a particular stage it'd pop its pipelines out of the dictionary until none were left.

Then you'd have `util.pipelines` be a well-known thing that is always cloned as part of the core compiler. If you have random floating attributes that's not possible.

The current situation with PreprocessingPassPipelineAttr in util, the pass in common, and the free floating attr is bad, and adding additional coupling now with binding layers and propagating the special "preprocessing" nature of things makes it worse.

If you need to get this in ASAP then special case the `"preprocessing_pipeline"` string in WrapEntryPoints and remove all the rest of the code so it's easier to clean this up in the future.

(it's my bad for not pushing back during initial review of the feature - I don't think I was paying enough attention - sorry about that)

MaheshRavishankar committed:
[Util] Move Attribute based pipeline to Util dialect.

The initial implementation of attribute-based pipelines was restricted
to preprocessing. This is now generalized to allow applying an
attribute based pipeline specification anywhere in the compiler.

The `util.pipelines` attribute is a `DictionaryAttr` that contains a
list of pipelines that can be applied anywhere. For example, currently
a `preprocessing` pipeline can be specified as follows,

```
util.pipelines = {
    "preprocessing" : #util.pipeline<"iree-preprocessing-make-single-dispatch">
}
```

Currently `iree-compile` only understands the "preprocessing" entry,
but more such entry points can be added in the future.

Signed-off-by: MaheshRavishankar <mahesh.ravishankar@gmail.com>
